quinn-proto/src/connection/mod.rs

```diff
@@ -402,67 +405,166 @@ where
+                    } else {
+                        0
+                    } as u64;
+                    debug_assert!(untracked_bytes <= self.path.mtu as u64);
+
+                    let bytes_to_send = u64::from(self.path.mtu) + untracked_bytes;
+
+                    if self.congestion_blocked(bytes_to_send) {
+                        space_idx += 1;
                         congestion_blocked = true;
+                        // TODO:Should those rather be break?
+                        // Maybe not due to loss probes
                         continue;
                     }
+
+                    // Check whether the next datagram is blocked by pacing
                     let smoothed_rtt = self.path.rtt.get();
-                    let window = self.path.congestion.window();
                     if let Some(delay) = self.path.pacing.delay(
                         smoothed_rtt,
-                        self.path.mtu as u64,
+                        bytes_to_send,
                         self.path.mtu,
-                        window,
+                        self.path.congestion.window(),
                         now,
                     ) {
                         self.timers.set(Timer::Pacing, delay);
+                        space_idx += 1;
                         congestion_blocked = true;
+                        // TODO:Should those rather be break?
+                        // Maybe not due to loss probes
                         continue;
                     }
                 }
+
+                // Finish current packet
+                if let Some(mut builder) = builder.take() {
+                    // Pad the packet to make it suitable for sending with GSO
+                    // which will always send the maximum PDU.
+                    builder.min_size =
+                        builder.datagram_start + (self.path.mtu as usize) - builder.tag_len;
+
+                    self.finish_and_track_packet(now, builder, sent_frames.take(), &mut buf);
+
+                    debug_assert_eq!(buf.len(), buf_capacity, "Packet must be padded");
+                }
+
+                // Allocate space for another datagram
+                buf_capacity += self.path.mtu as usize;
+                if buf.capacity() < buf_capacity {
+                    buf.reserve(buf_capacity - buf.capacity());
+                }
+                num_datagrams += 1;
+                coalesce = true;
+                min_datagram_size = 0;
+            } else {
+                // We can append/coalesce the next packet into the current
+                // datagram.
+                // Finish current packet without adding extra padding
+                if let Some(builder) = builder.take() {
+                    self.finish_and_track_packet(now, builder, sent_frames.take(), &mut buf);
+                }
             }
 
+            debug_assert!(buf_capacity - buf.len() >= MIN_PACKET_SPACE);
```

Comment:
Should this be `==` (or rather, `debug_assert_eq`)? We seem to be inferring the size limit for the current packet from `buf_capacity` in various places.

Reply:
Why should it be? If we are starting a new datagram we have lots of space. But the same branch is also taken in the coalescing case, when the current datagram already contains a packet

Reply:
Er, right.